crates/evm/src/execute.rs

```diff
@@ -14,33 +19,123 @@ pub use reth_storage_errors::provider::ProviderError;
 ///
 /// The trait cannot commit the state changes to the database directly, see [`EvmCommit`].
 pub trait EvmTransact {
-    /// The environment for the evm.
-    type Env;
-    /// The output produced by the evm.
-    type Output;
-    /// The error produced by the evm.
-    type Error;
+    /// The database used by the evm.
+    type DB: Database;
 
     /// Transacts with the current env to produce an output without
     /// committing to the underlying database.
-    fn transact(&mut self) -> Result<Self::Output, Self::Error>;
+    fn transact(&mut self) -> EVMResult<<<Self as EvmTransact>::DB as Database>::Error>;
 
     /// Returns a mutable reference to the current env.
-    fn env_mut(&mut self) -> &mut Self::Env;
+    fn env_mut(&mut self) -> &mut Env;
 
     /// Returns a reference to the current env.
-    fn env(&self) -> &Self::Env;
+    fn env(&self) -> &Env;
+
+    /// Returns the current env with handler cfg.
+    fn env_with_handler_cfg(&self) -> EnvWithHandlerCfg;
+
+    /// Returns a reference to the current database.
+    fn db(&self) -> &Self::DB;
+
+    /// Returns a mutable reference to the current database.
+    fn db_mut(&mut self) -> &mut Self::DB;
+
+    /// Returns the reference of transaction.
+    fn tx(&self) -> &TxEnv;
+
+    /// Returns the mutable reference of transaction.
+    fn tx_mut(&mut self) -> &mut TxEnv;
+
+    /// Returns the reference of block.
+    fn block(&self) -> &BlockEnv;
+}
+
+/// Trait that represents the output of a transaction and provides access to the state.
+pub trait EvmTransactOutput {
+    /// Returns the state produced by the transaction.
+    fn state(&self) -> EvmState;
 }
```

Comment:
Why is this needed?

Reply:
not needed for this iteration, just removed